xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,399 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package gracefulswitch implements a graceful switch load balancer.
+package gracefulswitch
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/resolver"
+)
+
+var (
+	errBalancerClosed = errors.New("gracefulSwitchBalancer is closed")
+	defaultState      = balancer.State{
+		ConnectivityState: connectivity.Connecting,
+		Picker:            base.NewErrPicker(balancer.ErrNoSubConnAvailable),
+	}
+)
+
+// NewGracefulSwitchBalancer returns a graceful switch Balancer.
+func NewGracefulSwitchBalancer(cc balancer.ClientConn, opts balancer.BuildOptions) *Balancer {
+	return &Balancer{
+		cc:    cc,
+		bOpts: opts,
+	}
+}
+
+// Balancer is a utility to gracefully switch from one balancer to
+// a new balancer.
+type Balancer struct {
+	bOpts balancer.BuildOptions
+	cc    balancer.ClientConn
+
+	mu              sync.Mutex // also guards all fields within balancerCurrent and balancerPending
+	balancerCurrent *balancerWrapper
+	balancerPending *balancerWrapper
+	closed          bool // set to true when this balancer is closed
+
+	// currentMu must be locked before mu. This mutex guards against this
+	// sequence of events: UpdateSubConnState() called, finds the
+	// balancerCurrent, gives up lock, updateState comes in, causes Close() on
+	// balancerCurrent before the UpdateSubConnState is called on the
+	// balancerCurrent.
+	currentMu sync.Mutex
+}
+
+// caller must hold gsb.mu.
+func (gsb *Balancer) updateState(bw *balancerWrapper, state balancer.State) {
```

Comment:
I think it would be cleaner to implement:

- On `*Balancer`: only methods called into by the channel + any helpers for those methods
- On `*balancerWrapper`: methods called into by the child + any helpers for those methods

I.e. move the body of this implementation (and the others like it) to `balancerWrapper.UpdateState` (etc) and delete it from here.

There's some fragmentation of implementation between the two places for these methods, and that makes the code harder to understand.  And sorting the implementations this way makes it much easier to see what direction the calls are coming from.  This also clears up a lot of the confusion about when mu needs to be held.